py/mpconfig.h: Define templates for "feature levels".

jimmo opened the pull request:
This is a proposal/RFC for a larger piece of work (related to #7538) to simplify enabling/disabling features.

My goals are:
 - Remove redundancy from mpconfigport.h (never set a value to the default -- make it clear exactly what's being enabled).
 - Improve consistency between ports. All "similar" ports (i.e. approx same flash size) should get the same features.
 - Simplify mpconfigport.h -- just get default/sensible options for the size of the port.
 - Make it easy for defining constrained boards (e.g. STM32F0/L0) can just set a lower level.
 
This PR just makes a step towards this and defines the "core" level as the current default feature set, and a "minimal" level to turn off everything (and updates just the minimal port to use it). It highlights a few things that are enabled on the minimal port that perhaps shouldn't be (~2kiB).

I've added a couple of placeholder "levels" as candidate suggestions for where the other ports will roughly land.

This is a no-op change (the x86/arm minimal binary is the same size and contains the exact same symbols). All other ports unchanged.

jimmo committed:
py/mpconfig.h: Define starting templates for "feature levels".

Defines the "core" level as the current default feature set, and "minimal"
to turn off everything.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

stinos commented on py/mpconfig.h:
```diff
@@ -62,6 +62,24 @@
+#define _MICROPY_CONFIG_ROM_LEVEL_COMMON (MICROPY_CONFIG_ROM_LEVEL >= 2)
+#define _MICROPY_CONFIG_ROM_LEVEL_CONVENIENCE (MICROPY_CONFIG_ROM_LEVEL >= 3)
+#define _MICROPY_CONFIG_ROM_LEVEL_ALL (MICROPY_CONFIG_ROM_LEVEL >= 4)
+#define _MICROPY_CONFIG_ROM_LEVEL_EVERYTHING (MICROPY_CONFIG_ROM_LEVEL >= 5)
```
'all' and 'everything' are quite similar; use 'coverage' for the latter?

Wrt 'rom level' naming: since it's a set of feature levels, perhaps just name it like that (MICROPY_FEATURE_LEVEL) as well?

I can imagine that instead of the single-line description of which level has what, there'd be more elaborate comments. In that case the explanation of which one does what is going to be further away from where it gets defined. And now you have to repeat the numbers. How about moving those comments here, together with their respective definition? I.e. 

```
 // Only enable core features (constrained flash, e.g. STM32L072)
 #define _MICROPY_CONFIG_ROM_LEVEL_CORE (MICROPY_CONFIG_ROM_LEVEL >= 1)
```

Perhaps overkill, but people might want to adopt this principle and define their own levels which are somewhere in between existing ones, which can be done by leaving gaps here i.e. 10, 20, 30, ...

jimmo replied:
> 'all' and 'everything' are quite similar; use 'coverage' for the latter?

Yep, I think you're right.

> Wrt 'rom level' naming: since it's a set of feature levels, perhaps just name it like that (MICROPY_FEATURE_LEVEL) as well?

I did want to also add RAM levels too (e.g. rp2 is ram but not flash constrained), but I think given that ROM generally maps to features, this makes sense. My reasoning for calling it ROM was to emphasize that we're "saving ROM", not "cutting features" (although I agree that's a fairly abstract distinction).

> Perhaps overkill, but people might want to adopt this principle and define their own levels which are somewhere in between existing ones, which can be done by leaving gaps here i.e. 10, 20, 30, ...

I guess so... perhaps this could be useful for board-specific overrides. I wanted to avoid having to decide "is this as 27 or 26 level feature discussions" in the future.

On that though, I think it's worth putting #defines for the levels (somewhere before mpconfigboard.h gets imported) so that the numbers themselves are not meaningful. (i.e. minimal/mpconfigport.h should set it to "MINIMAL" not "0").

jimmo committed:
ports/minimal: Use feature level minimal.

Update minimal port to use the new "minimal" rom level config (this is a
no-op change the binary is the same size and contains the exact same
symbols).

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
Updated the branch:
 - Renamed the levels to: minimal, core features, basic features, extra features, full features, everything.
 - Added defines for the levels (instead of using the numbers directly) and made them 0, 10, 20, etc.
 - The `AT_LEAST` macros are no longer private (mpconfigport.h might wish to use them, based on mpconfigboard.h setting the level).
 - Split the mpconfig change from the ports/minimal change.

I think it's still worth separating full from everything. Full would be a board with unlimited flash, but there might still be features that we might not enable that only make sense for coverage (extra debugging, etc). Can always remove it if we don't end up using it. Maybe when we add the RAM levels, it'll make sense that the "everything" level is combined.

peterhinch commented:
If this goes towards addressing https://github.com/micropython/micropython/issues/5564 then it's a big step forward.

jimmo commented:
Updated anything enabled via level=core that was dependent on a parent option also enabled via level=core to just be unconditionally enabled by default. i.e. `MICROPY_PY_SYS_EXIT` depends on `MICROPY_PY_SYS`.

dpgeorge commented:
Very nice, thank you!  I wanted to do this for a while now.

Merged in 01374d941f9d7398e35990b574769b20c6779457 and 910e060f93c02c51a498f57ef77ccbe02566855b
